logging: change uint32_t to log_timestamp_t for timestamps

valeriosetti opened the pull request:
The log_timestamp_t type get the proper value (uint32_t or uint64_t)
based on the platform and defconfig definitions. However there are
some variables or functions' parameters that are still forced to
uint32_t type.
This fix tries to makes this more consistent.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

nordic-krch commented on include/logging/log_output.h:
```diff
@@ -237,7 +237,7 @@ static inline void log_output_hostname_set(const struct log_output *output,
  *
  * @param freq Frequency in Hz.
  */
-void log_output_timestamp_freq_set(uint32_t freq);
```
this is frequency of the timestamp and not the timestamp. Uint32_t should be enough (>4GHz).

valeriosetti replied:
Yes, you're right. Besides, even in case (quite remote) we want to address a situation with frequency higher than 4GHz, the type name "log_timestamp_t" wouldn't be correct.
Sorry for this: I was looking at timestamps and then I ended up doing extra useless changes

nordic-krch commented on subsys/logging/log_output.c:
```diff
@@ -43,8 +43,8 @@ static const char *const colors[] = {
 
-static uint32_t freq;
-static uint32_t timestamp_div;
+static log_timestamp_t freq;
```
both variables should stay uint32_t since it's frequency and frequency divider.

valeriosetti replied:
Yes, of course the same holds also for this point. I'll revert it

nordic-krch commented on subsys/logging/log_output.c:
```diff
@@ -170,7 +170,7 @@ void log_output_flush(const struct log_output *output)
+			   uint32_t flags, log_timestamp_t timestamp)
 {
 	int length;
 	bool format =
```
could you add handling of 64 bit timestamp in `!format` case? Currently it always prints `%08lu`

nordic-krch replied:
formatting case needs more work as it is currently limited to 99 hours.

valeriosetti replied:
Yes, sure, that's something I can do!
I'll try to make it asap and re-push it again

nordic-krch commented on subsys/logging/log_output.c:
```diff
@@ -744,7 +744,7 @@ void log_output_dropped_process(const struct log_output *output, uint32_t cnt)
 }
 
-void log_output_timestamp_freq_set(uint32_t frequency)
+void log_output_timestamp_freq_set(log_timestamp_t frequency)
```
should stay uint32_t

nordic-krch reviewed (changes requested):
thanks for spotting this, few comments

valeriosetti committed:
logging: fix for 64 bit timestamp + new timestamp format option

Timestamps can be 32 or 64 bit long based on the platform and
build configurations. The proper way to handle these timestamps
is to use the ad-hoc log_timestamp_t variable.
This patch fixes some timestamp's reference which were still
using uint32_t changing them to log_timestamp_t.

Moreover also a new config is added in order to print the
timestamp as the Linux's kernel format. This might be useful
in AMP platforms in which Linux's and Zephyr's logs must be
interleaved in order to get a more comprehensive log solution.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
Hi @nordic-krch, 
I just pushed the updated code in which:
- I removed type changes to "frequency" variables, as you correctly pointed out
- I added the support to print in "syslogd style" even when CONFIG_NEWLIB_LIBC is not defined
- I added also a config to print the timestamp as the Linux's kernel does. I just used this in a project I worked on recently since the customer wanted to have the Zephyr's logs interleaved with the Linux's ones, so having the same format made further analysis processes easier

nordic-krch commented on subsys/logging/log_output.c:
```diff
@@ -216,14 +278,25 @@ static int timestamp_print(const struct log_output *output,
 					hours, mins, seconds, ms * 1000U + us);
 #endif
 		} else {
+#if !IS_ENABLED(CONFIG_LOG_OUTPUT_FORMAT_LINUX_TIMESTAMP)
```
i would recommend using 
```
if (!IS_ENABLED(CONFIG_LOG_OUTPUT_FORMAT_LINUX_TIMESTAMP)) {
}
```
as it seems to me that there might be some warnings with unused variables (e.g. `mins`) in certain configurations.

valeriosetti replied:
You're right: I fixed this one

nordic-krch commented on subsys/logging/log_output.c:
```diff
@@ -216,14 +278,25 @@ static int timestamp_print(const struct log_output *output,
 			length = print_formatted(output, "%s.%06uZ ",
 						 time_str, ms * 1000U + us);
 #else
```
imo, this could also be converted to `if (IS_ENABLED(..))` so that ifdef around `get_YMD_from_seconds()` implementation could be removed.

valeriosetti replied:
I tried to change this one as well, but of course removing the `#if defined` around the `strftime()` function causes a warning for "undefined function" when building without `CONFIG_NEWLIB_LIBC`.

My proposal in order to remove all the conditionals `#if def` was to add an `__attribute__((unused))` for the `get_YMD_from_seconds()` function. Is this acceptable?

nordic-krch reviewed (commented):
couple of suggestions regarding use if compiler check instead of ifdef.
